File: C++/stack.cpp
//
// Created by Daniel on 31.03.2015.
//
#include "stack.h"
#include <boost/algorithm/string.hpp>
#include <boost/lexical_cast.hpp>
#include <sstream>
using namespace boost::algorithm;

bool checkParentheses(std::string str) {
    auto parentheses = stack<char>();
    for (auto c : str) {
        if (c == ')' || c == ']' || c == '}' || c == '>') {
            if (parentheses.isEmpty() || parentheses.pop() != c)
                return false;
        }
        if (c == '(') parentheses.push(')');
        if (c == '[') parentheses.push(']');
        if (c == '{') parentheses.push('}');
        if (c == '<') parentheses.push('>');
    }
    return parentheses.isEmpty();
}

std::string toBinary(int num) {
    return toBase(num, 2);
}

std::string toBase(int num, int base) {
    auto digits = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ";
    auto digitsInBase = stack<char>();
    while (num > 0) {
        digitsInBase.push(digits[num % base]);
        num /= base;
    }

    auto res = std::string();

    while (!digitsInBase.isEmpty())
    {
        res += digitsInBase.pop();
    }

    return res;
}

std::string toPostfix(std::string str) {
    std::vector<std::string> tokens;
    split(tokens, str, boost::is_space(), token_compress_on);
    auto prec = [] (auto str) {
        if(str == "+" || str == "-")
            return 1;
        if(str == "*" || str == "/")
            return 2;
        return 0;
    };

    auto operators = stack<std::string>();

    std::stringstream res;

    for (auto token : tokens) {
        if(token == "(")
            operators.push(token);
        else if (token == ")") {
            auto op = operators.pop();
            while (op != "(") {
                res << op << " ";
                op = operators.pop();
            };
        }
        else if (token == "+" || token == "-" || token == "*" || token == "/") {
            while (!operators.isEmpty() && prec(token) <= prec(operators.peek()))
                res << operators.pop() << " ";
            operators.push(token);
        }
        else
            res << token << " ";
    }

    while (!operators.isEmpty())
        res << operators.pop() << " ";

    return res.str().erase(res.str().length() - 1);
}

int evalPostfix(std::string str) {
    std::vector<std::string> tokens;
    split(tokens, str, boost::is_space(), token_compress_on);
    auto evalStack = stack<int>();
    auto eval = [&](auto binOp) {
        auto right = evalStack.pop();
        auto left = evalStack.pop();
        evalStack.push(binOp(left, right));
    };

    for(auto token : tokens)
    {
        if(token == "+") eval([](auto x, auto y) { return x + y; });
        else if(token == "-") eval([](auto x, auto y) { return x - y; });
        else if(token == "*") eval([](auto x, auto y) { return x * y; });
        else if(token == "/") eval([](auto x, auto y) { return x / y; });
        else evalStack.push(boost::lexical_cast<int>(token));
    }

    return evalStack.pop();
}

File: Go/Stack/stack.go
package Stack
import (
    "fmt"
    "strings"
    "strconv"
)

type Stack struct {
    data []interface{}
}

func (s *Stack) size() int {
    return len(s.data)
}

func (s *Stack) push(item interface{}) {
    s.data = append(s.data, item)
}

func (s *Stack) peek() interface{} {
    return s.data[len(s.data) - 1]
}

func (s *Stack) pop() interface{} {
    idx := len(s.data) - 1
    res := s.data[idx]
    s.data = s.data[0:idx]
    return res
}

func (s *Stack) isEmpty() bool {
    return len(s.data) == 0
}

func checkParentheses(str string) bool {
    stack := new(Stack);
    for _, c := range str {
        if c == ')' || c == ']' || c == '}' || c == '>' {
            if stack.isEmpty() || stack.pop() != c {
                return false;
            }
        }
        if c == '(' { stack.push(')') }
        if c == '[' { stack.push(']') }
        if c == '{' { stack.push('}') }
        if c == '<' { stack.push('>') }
    }

    return stack.isEmpty()
}

func toBase(num int, base int) string {
    digits := "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ"
    stack := new(Stack)
    for num > 0 {
        stack.push(digits[num % base])
        num /= base
    }

    res := ""
    for !stack.isEmpty() {
        res += fmt.Sprintf("%c", stack.pop())
    }

    return res
}

func toBinary(num int) string {
    return toBase(num, 2);
}

func toPostfix(str string) string {
    tokens := strings.Split(str, " ")
    prec := func(token string) int {
        switch token {
            case "+", "-":
            return 1
            case "*", "/":
            return 2
        }
        return 0
    }

    operators := new(Stack)

    var res []string

    for _, token := range tokens {
        switch token {
            case "(":
            operators.push("(")
            case ")":
            op := operators.pop()
            for op != "(" {
                res = append(res, op.(string))
                op = operators.pop()
            }
            case "+", "-", "*", "/":
            for !operators.isEmpty() && prec(token) <= prec(operators.peek().(string)) {
                res = append(res, operators.pop().(string))
            }
            operators.push(token)
            default:
            res = append(res, token)
        }
    }

    for !operators.isEmpty() {
        res = append(res, operators.pop().(string))
    }

    return strings.Join(res, " ")
}

type binOp func(int, int) int

func evalPostfix(str string) int {
    tokens := strings.Split(str, " ")
    evalStack := new(Stack)
    eval := func(f binOp) {
        right := evalStack.pop()
        left := evalStack.pop()
        evalStack.push(f(left.(int), right.(int)))
    }

    for _, token := range tokens {
        switch token {
            case "+":
            eval(func(x int, y int) int {return x + y})
            case "-":
            eval(func(x int, y int) int {return x - y})
            case "*":
            eval(func(x int, y int) int {return x * y})
            case "/":
            eval(func(x int, y int) int {return x / y})
            default:
            num, _ := strconv.ParseInt(token, 0, 32)
            evalStack.push(int(num))
        }
    }
    return evalStack.pop().(int)
}